File: getsploit.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import json
import vulners
import re
import os
import six
import texttable
import urllib
from six import string_types
from clint.textui import progress
from os.path import expanduser
import errno
import hashlib
import getpass

import redis

if six.PY2:
    import argparse
else:
    from optparse import OptionParser as argparse

if six.PY3:
    unicode_type = str
    bytes_type = bytes
else:
    unicode_type = unicode
    bytes_type = str


SCRIPTNAME = os.path.split(os.path.abspath(__file__))
DBPATH = os.path.join(expanduser("~"), '.getsploit')
DBFILE = os.path.join(DBPATH, 'getsploit.db')
KEYFILE = os.path.join(DBPATH, 'vulners.key')
MAX_ATTEMPTS = 3

if not os.path.exists(DBPATH):
    try:
        os.makedirs(DBPATH)
    except OSError as exception:
        if exception.errno != errno.EEXIST:
            raise

try:
    LOCAL_SEARCH_AVAILABLE = True
    redis_host = 'localhost'
    redis_port = 6379
    redis_password = None
    redis_db = 0
    redis_url = os.environ.get("REDIS_URL")
    if redis_url:
        r = redis.from_url(redis_url)
    else:
        r = redis.StrictRedis(
            host=redis_host, port=redis_port, password=redis_password, db=redis_db)
    r.ping()
except Exception as e:
    print('Redis connection error: ', e)
    LOCAL_SEARCH_AVAILABLE = False


class RedisClient(vulners.Vulners):

    api_endpoints = {
        'search': "/api/v3/search/lucene/",
        'software': "/api/v3/burp/software/",
        'apiKey': "/api/v3/apiKey/valid/",
        'searchsploitdb': "/api/v3/archive/getsploit/"
    }

    def __init__(self, api_key=None):
        super(RedisClient, self).__init__(api_key=api_key)

    def check_api_key(self):
        response = self._Vulners__opener.get(self.vulners_urls['apiKey'])
        return json.loads(response.read().decode())

    def search_local(self, query, lookup_fields=None, limit=500, offset=0, fields=None):
        searchQuery, dataDocs = self.searchExploit(query, lookup_fields, limit, offset, fields)
        return dataDocs

    def search(self, query, lookup_fields=None, limit=500, offset=0, fields=None):
        searchQuery, dataDocs = self.searchExploit(query, lookup_fields, limit, offset, fields)
        return dataDocs

    def download_getsploit_db(self):
        """
        Downloads the getsploit exploit database from Vulners API.

        :return: The getsploit exploit data.
        :rtype: list of dict
        """
        if not LOCAL_SEARCH_AVAILABLE:
            raise Exception("Local search is not available")
        print("Downloading getsploit database archive. Please wait, it may take time. Usually around 5-10 minutes.")
        download_request = urllib.request.urlopen(self.vulners_urls['searchsploitdb'])
        data = download_request.read()
        # Store the data using SHA-512 as key
        key = hashlib.sha512(data).hexdigest()
        r.set(key, data)
        archive = zipfile.ZipFile(data)
        archive.extractall(expanduser("~/.getsploit"))
        archive.close()
        with open(expanduser("~/.getsploit/getsploit.json"), "r") as f:
            return json.loads(f.read())

def authenticate():
    """
    Authenticates the client using the provided API key.

    :return: The authenticated client.
    :rtype: RedisClient
    """
    api_key_file = expanduser("~/.getsploit/apikey.txt")
    if not os.path.exists(api_key_file):
        api_key = getpass.getpass("Enter Vulners API key: ")
        with open(api_key_file, "w") as f:
            f.write(hashlib.sha512(api_key.encode("utf-8")).hexdigest())
    else:
        with open(api_key_file, "r") as f:
            saved_api_key_hash = f.read().strip()
        for i in range(3):
            api_key = getpass.getpass("Enter Vulners API key: ")
            if hashlib.sha512(api_key.encode("utf-8")).hexdigest() == saved_api_key_hash:
                break
            else:
                print("Invalid API key")
        else:
            print("Too many failed attempts")
            sys.exit(1)
    return RedisClient(api_key)

def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(description="Exploit search and download utility")
    parser.add_argument("query", metavar="query", type=str, nargs="+",
                        help="Exploit search query. See https://vulners.com/help for the detailed manual.")
    parser.add_argument("-t", "--title", action="store_true",
                        help="Search JUST the exploit title (Default is description and source code).")
    parser.add_argument("-j", "--json", action="store_true",
                        help="Show result in JSON format.")
    parser.add_argument("-m", "--mirror", action="store_true",
                        help="Mirror (aka copies) search result exploit files to the subdirectory with your search query name.")
    parser.add_argument("-c", "--count", nargs=1, type=int, default=10,
                        help="Search limit. Default 10.")
    parser.add_argument("-l", "--local", action="store_true",
                        help="Perform search in the local database instead of searching online.")
    parser.add_argument("-u", "--update", action="store_true",
                        help="Update local database with latest information from Vulners API.")
    args = parser.parse_args()

    # Check if search query is provided
    if not args.query:
        print("No search query provided. Type software name and version to find exploit.")
        sys.exit(1)

    # Authenticate the client
    client = authenticate()

    # Update the database if requested
    if args.update:
        if not LOCAL_SEARCH_AVAILABLE:
            print("Local search is not available")
            sys.exit(1)
        try:
            searchsploit_data = client.download_getsploit_db()
            for data in searchsploit_data:
                query = "bulletinFamily:exploitdb AND id:{} AND type:exploitdb".format(data["id"])
                response = client.search(query)
                if response["total"] == 1:
                    doc = response["search"][0]
                    doc["sourceData"] = data["sourceData"]
                    client.insert(doc, "exploitdb")
                else:
                    data.pop("sourceData")
                    client.insert(data, "exploitdb")
            print("Database updated")
        except urllib.error.URLError:
            print("Could not download getsploit database. Please check your internet connection and try again.")
        sys.exit(0)

    # Search the database
    if args.local:
        search_results = client.search_local(args.query[0], fields=["title"] if args.title else None, limit=args.count)
    else:
        search_results = client.search(args.query[0], fields=["title"] if args.title else None, limit=args.count)

    # Display results
    if args.json:
        print(json.dumps(search_results))
    else:
        table = Texttable()
        table.set_cols_dtype(["t", "t", "t"])
        table.set_cols_align(["c", "l", "c"])
        table.set_cols_width(["20", "30", "100"])
        table.add_rows([["ID", "Exploit Title", "URL"]] + [[res["id"], res["title"], res["vhref"]] for res in search_results])
        print(table.draw())

File: cmd/getsploit.go
package main

import (
    "crypto/sha512"
    "encoding/json"
    "fmt"
    "io/ioutil"
    "os"
    "path/filepath"
    "strings"

    "github.com/howeyc/gopass"
    "github.com/fatih/color"
)

const (
    dbPath = "~/.getsploit"
    maxAttempts = 3
)

type VulnersClient struct {
    APIKey string
}

type RedisClient struct {
    VulnersClient
}

func (c *RedisClient) Search(query string, fields []string, limit int, offset int) (map[string]interface{}, error) {
    response, err := c._get(fmt.Sprintf("/api/v3/search/lucene/?query=%s&fields=%s&size=%d&sort=order:desc", query, strings.Join(fields, ","), limit), nil)
    if err != nil {
        return nil, err
    }
    var data map[string]interface{}
    if err := json.Unmarshal(response, &data); err != nil {
        return nil, err
    }
    return data, nil
}

func (c *RedisClient) _get(endpoint string, params map[string]string) ([]byte, error) {
    req, err := c.newRequest("GET", endpoint, nil)
    if err != nil {
        return nil, err
    }
    return c.do(req)
}

func (c *RedisClient) newRequest(method, endpoint string, data map[string]string) (*http.Request, error) {
    url := c.vulnersURL(endpoint)
    req, err := http.NewRequest(method, url, nil)
    if err != nil {
        return nil, err
    }
    req.Header.Set("User-Agent", "getsploit")
    req.Header.Set("X-Vulners-API-Key", c.APIKey)
    return req, nil
}

func (c *RedisClient) do(req *http.Request) ([]byte, error) {
    resp, err := c.HTTPClient.Do(req)
    if err != nil {
        return nil, err
    }
    defer resp.Body.Close()
    content, err := ioutil.ReadAll(resp.Body)
    if err != nil {
        return nil, err
    }
    return content, nil
}

func (c *RedisClient) vulnersURL(endpoint string) string {
    return fmt.Sprintf("https://vulners.com%s", endpoint)
}

func (c *RedisClient) authenticate() error {
    // Check if there's an API key
    apiKeyFile := filepath.Join(dbPath, "apikey.txt")
    apiKeyHash := ""
    if _, err := os.Stat(apiKeyFile); err == nil {
        content, err := ioutil.ReadFile(apiKeyFile)
        if err != nil {
            return err
        }
        apiKeyHash = strings.TrimSpace(string(content))
    }
    for i := 0; i < maxAttempts; i++ {
        fmt.Print("Enter Vulners API key: ")
        apiKey, err := gopass.GetPasswd()
        if err != nil {
            return err
        }
        inputHash := fmt.Sprintf("%x", sha512.Sum512(apiKey))
        if apiKeyHash == inputHash {
            break
        } else {
            color.Red("Invalid API key")
        }
    }
    c.APIKey = string(apiKey)
    return nil
}

func main() {
    // Parse command line arguments
    if len(os.Args) == 1 {
        fmt.Println("No search query provided. Type software name and version to find exploit.")
        os.Exit(1)
    }
    query := os.Args[1]
    fields := []string{"description", "sourceData"}
    limit := 10
    titleOnly := false
    mirrorFiles := false
    localSearch := false
    updateDB := false
    for _, arg := range os.Args[2:] {
        switch arg {
        case "-t", "--title":
            fields = []string{"title"}
            titleOnly = true
        case "-j", "--json":
            // not implemented
            fmt.Println("JSON output not yet implemented")
            os.Exit(1)
        case "-m", "--mirror":
            mirrorFiles = true
        case "-c", "--count":
            if len(os.Args) < 4 {
                fmt.Println("Error: no count provided")
                os.Exit(1)
            }
            var err error
            limit, err = strconv.Atoi(os.Args[3])
            if err != nil {
                fmt.Println("Error: invalid count")
                os.Exit(1)
            }
        case "-l", "--local":
            localSearch = true
        case "-u", "--update":
            updateDB = true
        }
    }

    // Check if search query is provided
    if query == "" {
        fmt.Println("No search query provided. Type software name and version to find exploit.")
        os.Exit(1)
    }

    // Initialize Vulners client
    client := &RedisClient{VulnersClient{}}
    if err := client.authenticate(); err != nil {
        fmt.Printf("Error: %v\n", err)
        os.Exit(1)
    }

    // Perform search
    var searchResults []interface{}
    if localSearch {
        dataDocs, err := client.SearchLocal(query, []string{"title"}, limit, 0, fields)
        if err != nil {
            fmt.Printf("Error: %v\n", err)
            os.Exit(1)
        }
        for _, doc := range dataDocs {
            searchResults = append(searchResults, doc)
        }
    } else {
        data, err := client.Search(query, fields, limit, 0)
        if err != nil {
            fmt.Printf("Error: %v\n", err)
            os.Exit(1)
        }
        searchResults = data["data"].([]interface{})
    }

    // Display results
    table := tablewriter.NewWriter(os.Stdout)
    table.SetHeader([]string{"ID", "Exploit Title", "URL"})
    for _, result := range searchResults {
        resultMap := result.(map[string]interface{})
        title := resultMap["title"].(string)
        if titleOnly {
            if strings.Contains(strings.ToLower(title), strings.ToLower(query)) {
                table.Append([]string{
                    fmt.Sprintf("%.0f", resultMap["id"].(float64)),
                    title,
                    fmt.Sprintf("https://vulners.com/#!exploitdb/view/%.0f", resultMap["id"].(float64)),
                })
            }
        } else {
            table.Append([]string{
                fmt.Sprintf("%.0f", resultMap["id"].(float64)),
                title,
                fmt.Sprintf("https://vulners.com/#!exploitdb/view/%.0f", resultMap["id"].(float64)),
            })
        }
    }
    table.Render()

    // Mirror files
    if mirrorFiles {
        for _, result := range searchResults {
            resultMap := result.(map[string]interface{})
			if len(resultMap["sourceData"].([]interface{})) > 0 {
                fileName := fmt.Sprintf("%.0f_%s", resultMap["id"].(float64), strings.ReplaceAll(resultMap["title"].(string), " ", "_"))
                filePath := fmt.Sprintf("./%s/%s", query, fileName)
                sourceURL := resultMap["sourceData"].([]interface{})[0].(map[string]interface{})["url"].(string)
                fileData, err := client.downloadFile(sourceURL)
                if err != nil {
                    fmt.Printf("Error downloading file %s: %v\n", sourceURL, err)
                    continue
                }
                err = ioutil.WriteFile(filePath, fileData, 0644)
                if err != nil {
                    fmt.Printf("Error writing file %s: %v\n", filePath, err)
                    continue
                }
                fmt.Printf("Downloaded %s\n", fileName)
            }
        }
    }

    // Update local database
    if updateDB {
        if err := client.updateLocalDB(); err != nil {
            fmt.Printf("Error updating local database: %v\n", err)
            os.Exit(1)
        }
        fmt.Println("Local database updated successfully")
    }
}


